api/src/main/java/org/openmrs/api/impl/ConditionServiceImpl.java

```diff
@@ -0,0 +1,142 @@
+ */
+package org.openmrs.api.impl;
+
+import org.openmrs.Condition;
+import org.openmrs.Patient;
+import org.openmrs.api.APIException;
+import org.openmrs.api.ConditionService;
+import org.openmrs.api.context.Context;
+import org.openmrs.api.db.ConditionDAO;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.Date;
+import java.util.List;
+
+/**
+ * This class implements the {@link ConditionService} interface
+ * It defines the methods to handle the condition domain object
+ *
+ * @since 2.2
+ */
+public class ConditionServiceImpl extends BaseOpenmrsService implements ConditionService{
+
+	private ConditionDAO conditionDAO;
+
+	public void setConditionDAO(ConditionDAO conditionDAO) {
+		this.conditionDAO = conditionDAO;
+	}
+
+	public ConditionDAO getConditionDAO() {
+		return conditionDAO;
+	}
+
+	/**
+	 * Saves a condition
+	 *
+	 * @param condition - the condition to be saved
+	 */
+	@Override
+	@Transactional
+	public Condition saveCondition(Condition condition) throws APIException {
+		return conditionDAO.saveCondition(condition);
+	}
+
+	/**
+	 * Gets a condition based on the uuid
+	 *
+	 * @param uuid - uuid of the condition to be returned
+	 * @return the condition
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public Condition getConditionByUuid(String uuid) {
+		return conditionDAO.getConditionByUuid(uuid);
+	}
+
+	/**
+	 * Gets a condition by id
+	 *
+	 * @param conditionId - the id of the Condition to retrieve
+	 * @return the Condition with the given id, or null if none exists
+	 * @throws APIException
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public Condition getCondition(Integer conditionId) throws APIException {
+		return conditionDAO.getCondition(conditionId);
+	}
+
+	/**
+	 * Gets a patient's active conditions
+	 *
+	 * @param patient - the patient to retrieve conditions for
+	 * @return a list of the patient's active conditions
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public List<Condition> getActiveConditions(Patient patient) {
+		return conditionDAO.getActiveConditions(patient);
+	}
+
+	/**
+	 * Voids a condition
+	 *
+	 * @param condition  - the condition to be voided
+	 * @param voidReason - the reason for voiding the condition
+	 */
+	@Override
+	@Transactional
+	public Condition voidCondition(Condition condition, String voidReason) {
+		condition.setVoided(true);
+		condition.setVoidedBy(Context.getAuthenticatedUser());
+		condition.setVoidReason(voidReason);
+		if(condition.getDateVoided() == null){
+			condition.setDateVoided(new Date());
+		}
+		Context.getConditionService().saveCondition(condition);
+		return condition;
+	}
+
+	/**
+	 * Revive a condition
+	 *
+	 * @param condition Condition to unvoid
+	 * @throws APIException
+	 * @should unset voided bit on given condition
+	 */
+	@Override
+	@Transactional
+	public Condition unvoidCondition(Condition condition) throws APIException {
+		condition.setVoided(false);
+		condition.setVoidedBy(null);
```

Comment:
Did you look at the ObsServiceImpl's unvoidObs()?

Reply:
I did. However, I also carefully studied EncounterServiceImpl's unvoidEncounter() .

Reply:
EncounterServiceImpl is a bit complicated for this case

Reply:
I have some reservations with the ObsServiceImpl voidObs() and unvoidObs(). This is because they both practically just save the Obs.